staging/src/k8s.io/client-go/util/csaupgrade/upgrade.go

```diff
@@ -107,21 +134,76 @@ func UpgradeManagedFields(
 			entry.Subresource == "")
 	})
 
-	// Wipe out last-applied-configuration annotation if it exists
-	annotations := accessor.GetAnnotations()
-	delete(annotations, csaAnnotationName)
+	return filteredManagers, nil
+}
 
-	// Commit changes to object
-	accessor.SetAnnotations(annotations)
-	accessor.SetManagedFields(filteredManagers)
+// Ensures a given manager exists that owns the given field paths
+// If there is an entry for the given manager already, inserts
+//
+//	field path into its ownership set.
+//
+// Otherwise, creates manager at latest APIVersion which owns the field
+// Returns an error if there was an issue with decoding/encoding. Or nil if
+// the given field path are now owned by the desired manager
+func EnsureFieldOwnership(
+	managedFields []metav1.ManagedFieldsEntry,
+	apiVersion string,
+	operation metav1.ManagedFieldsOperationType,
+	managerName string,
+	subresource string,
+	paths *fieldpath.Set,
```

Comment:
Will the paths depend on the apiVersion? They probably should. Since you're already passing this as a param, that's nice, but I would put them together since they are coupled.